pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Schema.java

```
@@ -74,6 +74,10 @@ default void validate(byte[] message) {
      */
     byte[] encode(T message);
 
+    default EncodeData encode(String topic, T message) {
```

There is a case that `encode` could fail due to a failure happened at the external schema registry. For example, if the message failed to register it's schema, `encode` could throw an exception, which will eventually be propagated to `sendAsync`.

From the perspective of users, they will see `sendAsync()` throwing an exception, rather than a failed future being returned by `sendAsync`.

It's better to change the signature to

```java
    default EncodeData encode(String topic, T message) throws PulsarClientException {
```

and catch the exception in `sendAsync`

```java
    public CompletableFuture<MessageId> sendAsync() {
        Message<T> message;
        try {
            message = getMessage();
        } catch (PulsarClientException e) {
            return CompletableFuture.failedFuture(e);
        }
```

---

There is the same issue with the `decode` method

---

+1

We can throw the `SchemaSerializationException`.

---

We should add a description of the `SchemaSerializationException` to the method comment; otherwise, we need to modify all encode and decode methods and their callers. The encode method already has a throws comment, so it may throw exceptions.